In [1]:
from sklearn.datasets import load_digits, load_breast_cancer
from sklearn.naive_bayes import BernoulliNB, MultinomialNB, GaussianNB
from sklearn.model_selection import cross_val_score

In [2]:
beast_cancer = load_breast_cancer()
digits = load_digits()

In [3]:
digits['data']

array([[  0.,   0.,   5., ...,   0.,   0.,   0.],
       [  0.,   0.,   0., ...,  10.,   0.,   0.],
       [  0.,   0.,   0., ...,  16.,   9.,   0.],
       ..., 
       [  0.,   0.,   1., ...,   6.,   0.,   0.],
       [  0.,   0.,   2., ...,  12.,   0.,   0.],
       [  0.,   0.,  10., ...,  12.,   1.,   0.]])

In [4]:
beast_cancer['data']

array([[  1.79900000e+01,   1.03800000e+01,   1.22800000e+02, ...,
          2.65400000e-01,   4.60100000e-01,   1.18900000e-01],
       [  2.05700000e+01,   1.77700000e+01,   1.32900000e+02, ...,
          1.86000000e-01,   2.75000000e-01,   8.90200000e-02],
       [  1.96900000e+01,   2.12500000e+01,   1.30000000e+02, ...,
          2.43000000e-01,   3.61300000e-01,   8.75800000e-02],
       ..., 
       [  1.66000000e+01,   2.80800000e+01,   1.08300000e+02, ...,
          1.41800000e-01,   2.21800000e-01,   7.82000000e-02],
       [  2.06000000e+01,   2.93300000e+01,   1.40100000e+02, ...,
          2.65000000e-01,   4.08700000e-01,   1.24000000e-01],
       [  7.76000000e+00,   2.45400000e+01,   4.79200000e+01, ...,
          0.00000000e+00,   2.87100000e-01,   7.03900000e-02]])

In [5]:
estimators = BernoulliNB(), MultinomialNB(), GaussianNB()
estimators_names = "BernoulliNB", "MultinomialNB", "GaussianNB"
datasets = beast_cancer, digits
datasets_names = "beast_cancer", "digits"

for estimator, estimator_name in zip(estimators, estimators_names):
    for data, data_name in zip(datasets, datasets_names):
        print("Accuracy for {} by {} is {}".format(
                data_name, estimator_name,
                cross_val_score(estimator, data['data'], data['target'], cv=5).mean()
            )
        )

Accuracy for beast_cancer by BernoulliNB is 0.627425933051
Accuracy for digits by BernoulliNB is 0.82477104598
Accuracy for beast_cancer by MultinomialNB is 0.89637552905
Accuracy for digits by MultinomialNB is 0.871470302548
Accuracy for beast_cancer by GaussianNB is 0.940377068103
Accuracy for digits by GaussianNB is 0.806520755555


Для вещественной выборки лучше срабоал классификатор GaussianNB, а для целочисленной MultinomialNB.

Максимальное качество классификации для датасета breast_cancer составило $94\%$ (GaussianNB).

Максимальное качество классификации для датасета digits составило $87\%$ (MultinomialNB).

Верны утвержения (с) и (d)